### PR TITLE
ISM Setup for History and Sessions up and running

### PR DESCRIPTION
**Pull request by chelma**
## Description
* Added a Lambda to set up ISM policies on the OpenSearch Domain
* Added code to set up the user-history and sessions policy inside that Lambda
* Added CDK and Python code to have the Lambda be event-triggered as part of the `create-cluster` command

## Tasks
* https://github.com/arkime/aws-aio/issues/63

## Testing
* Unit testing
* Ran the `create-cluster` workflow to test the History ISM setup process.  Specifically:
  * Ran the `create-cluster` workflow to setup the ISM policy initially
  * Manually deleted the ISM policy using CURL from the capture node Fargate container
```root@983eeb1aa43f:/# OPENSEARCH_PASS=$(aws secretsmanager get-secret-value --secret-id $OPENSEARCH_SECRET_ARN --output text --query SecretString)

root@983eeb1aa43f:/# curl -u "admin:${OPENSEARCH_PASS}" -X GET "https://${OPENSEARCH_ENDPOINT}:443/_plugins/_ism/policies/history"
{"_id":"history","_version":3,"_seq_no":3833,"_primary_term":1,"policy":{"policy_id":"history","description":"Delete Arkime history indices","last_updated_time":1687375102621,"schema_version":17,"error_notification":null,"default_state":"warm","states":[{"name":"warm","actions":[],"transitions":[{"state_name":"delete","conditions":{"min_index_age":"120d"}}]},{"name":"delete","actions":[{"retry":{"count":3,"backoff":"exponential","delay":"1m"},"delete":{}}],"transitions":[]}],"ism_template":[{"index_patterns":["history_v*"],"priority":95,"last_updated_time":1687375102621}]}}

root@983eeb1aa43f:/# curl -u "admin:${OPENSEARCH_PASS}" -X DELETE "https://${OPENSEARCH_ENDPOINT}:443/_plugins/_ism/policies/history"
{"_index":".opendistro-ism-config","_id":"history","_version":4,"result":"deleted","forced_refresh":true,"_shards":{"total":2,"successful":2,"failed":0},"_seq_no":3842,"_primary_term":1}

root@983eeb1aa43f:/# curl -u "admin:${OPENSEARCH_PASS}" -X GET "https://${OPENSEARCH_ENDPOINT}:443/_plugins/_ism/policies/history"
{"error":{"root_cause":[{"type":"status_exception","reason":"Policy not found"}],"type":"status_exception","reason":"Policy not found"},"status":404}
```
  * Ran the `create-cluster` workflow again to re-create the policy
  * Manually CURL'd the OpenSearch REST API to ensure it the policy was recreated

```
root@983eeb1aa43f:/# curl -u "admin:${OPENSEARCH_PASS}" -X GET "https://${OPENSEARCH_ENDPOINT}:443/_plugins/_ism/policies/history"
{"_id":"history","_version":1,"_seq_no":3843,"_primary_term":1,"policy":{"policy_id":"history","description":"Delete Arkime history indices","last_updated_time":1687375658942,"schema_version":17,"error_notification":null,"default_state":"warm","states":[{"name":"warm","actions":[],"transitions":[{"state_name":"delete","conditions":{"min_index_age":"120d"}}]},{"name":"delete","actions":[{"retry":{"count":3,"backoff":"exponential","delay":"1m"},"delete":{}}],"transitions":[]}],"ism_template":[{"index_patterns":["history_v*"],"priority":95,"last_updated_time":1687375658942}]}}
```
* Ran the `create-cluster` workflow to test the Sessions ISM setup process.  Specifically:
  * Ran the `create-cluster` workflow to setup the ISM policy initially
  * Manually deleted the ISM policy using CURL from the capture node Fargate container
```root@983eeb1aa43f:/# OPENSEARCH_PASS=$(aws secretsmanager get-secret-value --secret-id $OPENSEARCH_SECRET_ARN --output text --query SecretString)

root@983eeb1aa43f:/# curl -u "admin:${OPENSEARCH_PASS}" -X GET "https://${OPENSEARCH_ENDPOINT}:443/_plugins/_ism/policies/sessions"
{"_id":"sessions","_version":1,"_seq_no":4061,"_primary_term":1,"policy":{"policy_id":"sessions","description":"Arkime sessions3 Policy","last_updated_time":1687449005102,"schema_version":17,"error_notification":null,"default_state":"hot","states":[{"name":"hot","actions":[],"transitions":[{"state_name":"warm","conditions":{"min_index_age":"30d"}}]},{"name":"warm","actions":[{"retry":{"count":3,"backoff":"exponential","delay":"1m"},"force_merge":{"max_num_segments":1}},{"retry":{"count":3,"backoff":"exponential","delay":"1m"},"allocation":{"require":{"molochtype":"warm"},"include":{},"exclude":{},"wait_for":true}},{"retry":{"count":3,"backoff":"exponential","delay":"1m"},"replica_count":{"number_of_replicas":1}}],"transitions":[{"state_name":"delete","conditions":{"min_index_age":"30d"}}]},{"name":"delete","actions":[{"retry":{"count":3,"backoff":"exponential","delay":"1m"},"delete":{}}],"transitions":[]}],"ism_template":[{"index_patterns":["sessions3-*"],"priority":95,"last_updated_time":1687449005102}]}}

root@983eeb1aa43f:/# curl -u "admin:${OPENSEARCH_PASS}" -X DELETE "https://${OPENSEARCH_ENDPOINT}:443/_plugins/_ism/policies/sessions"
{"_index":".opendistro-ism-config","_id":"sessions","_version":2,"result":"deleted","forced_refresh":true,"_shards":{"total":2,"successful":2,"failed":0},"_seq_no":4062,"_primary_term":1}

root@983eeb1aa43f:/# curl -u "admin:${OPENSEARCH_PASS}" -X GET "https://${OPENSEARCH_ENDPOINT}:443/_plugins/_ism/policies/sessions"
{"error":{"root_cause":[{"type":"status_exception","reason":"Policy not found"}],"type":"status_exception","reason":"Policy not found"},"status":404}
```
  * Ran the `create-cluster` workflow again to re-create the policy
  * Manually CURL'd the OpenSearch REST API to ensure it the policy was recreated
```
root@983eeb1aa43f:/# curl -u "admin:${OPENSEARCH_PASS}" -X GET "https://${OPENSEARCH_ENDPOINT}:443/_plugins/_ism/policies/sessions"
{"_id":"sessions","_version":1,"_seq_no":4063,"_primary_term":1,"policy":{"policy_id":"sessions","description":"Arkime sessions3 Policy","last_updated_time":1687449457792,"schema_version":17,"error_notification":null,"default_state":"hot","states":[{"name":"hot","actions":[],"transitions":[{"state_name":"warm","conditions":{"min_index_age":"30d"}}]},{"name":"warm","actions":[{"retry":{"count":3,"backoff":"exponential","delay":"1m"},"force_merge":{"max_num_segments":1}},{"retry":{"count":3,"backoff":"exponential","delay":"1m"},"allocation":{"require":{"molochtype":"warm"},"include":{},"exclude":{},"wait_for":true}},{"retry":{"count":3,"backoff":"exponential","delay":"1m"},"replica_count":{"number_of_replicas":1}}],"transitions":[{"state_name":"delete","conditions":{"min_index_age":"30d"}}]},{"name":"delete","actions":[{"retry":{"count":3,"backoff":"exponential","delay":"1m"},"delete":{}}],"transitions":[]}],"ism_template":[{"index_patterns":["sessions3-*"],"priority":95,"last_updated_time":1687449457792}]}}
```

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: manage_arkime/aws_interactions/cloudwatch_interactions.py [hunk 1]
```diff
@@ -38,6 +38,55 @@ def __str__(self) -> str:
...
+class ConfigureIsmEventMetrics(ArkimeEventMetric):
+    def __init__(self, cluster_name: str, outcome: ConfigureIsmEventOutcome):
+        super().__init__()
+
+        self.cluster_name = cluster_name
+        self.event_type = constants.EVENT_DETAIL_TYPE_CREATE_ENI_MIRROR
+
+        self.value_success = 0
+        self.value_abort_exists = 0
+        self.value_abort_eni_type = 0
```
Comment (chelma) [inline manage_arkime/aws_interactions/cloudwatch_interactions.py:55]: Delete these lines; copy/paste leftover

File: manage_arkime/aws_interactions/cloudwatch_interactions.py [hunk 2]
```diff
@@ -38,6 +38,55 @@ def __str__(self) -> str:
...
+class ConfigureIsmEventOutcome(Enum):
+    SUCCESS="Success"
+    FAILURE="Failure"
+
+class ConfigureIsmEventMetrics(ArkimeEventMetric):
+    def __init__(self, cluster_name: str, outcome: ConfigureIsmEventOutcome):
+        super().__init__()
+
+        self.cluster_name = cluster_name
+        self.event_type = constants.EVENT_DETAIL_TYPE_CREATE_ENI_MIRROR
```
Comment (chelma) [inline manage_arkime/aws_interactions/cloudwatch_interactions.py:51]: Copy/paste error; set to correct event type

File: test_manage_arkime/lambda_configure_ism/test_configure_ism_handler.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+from lambda_configure_ism.configure_ism_handler import ConfigureIsmHandler
+from opensearch_interactions.opensearch_client import OpenSearchClient
+import aws_interactions.cloudwatch_interactions as cwi
+import constants as constants
+
+@mock.patch("lambda_configure_ism.configure_ism_handler.os")
+@mock.patch("lambda_configure_ism.configure_ism_handler.ism.setup_user_history_ism")
+@mock.patch("lambda_configure_ism.configure_ism_handler.AwsClientProvider")
+@mock.patch("lambda_configure_ism.configure_ism_handler.cwi")
+def test_WHEN_CreateEniMirrorHandler_handle_called_THEN_sets_up_mirroring(mock_cwi, mock_provider, mock_setup_history, mock_os):
```
Comment (chelma) [inline test_manage_arkime/lambda_configure_ism/test_configure_ism_handler.py:13]: Copy/paste error - rename

File: test_manage_arkime/lambda_configure_ism/test_configure_ism_handler.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            mock.ANY
+        ),
+    ]
+    assert expected_put_metrics_calls == mock_cwi.put_event_metrics.call_args_list
+
+# @mock.patch("lambda_create_eni_mirror.create_eni_mirror_handler.AwsClientProvider", mock.Mock())
+# @mock.patch("lambda_create_eni_mirror.create_eni_mirror_handler.cwi")
+# @mock.patch("lambda_create_eni_mirror.create_eni_mirror_handler.ssm_ops")
+# @mock.patch("lambda_create_eni_mirror.create_eni_mirror_handler.ec2i")
+# def test_WHEN_CreateEniMirrorHandler_handle_called_AND_unhandled_ex_THEN_handles_gracefully(mock_ec2i, mock_ssm_ops, mock_cwi):
```
Comment (chelma) [inline test_manage_arkime/lambda_configure_ism/test_configure_ism_handler.py:71]: Implement

File: manage_arkime/opensearch_interactions/ism_interactions.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    else:
+        # Create the policy
+        client.create_ism_policy(policies.ISM_ID_HISTORY, policy)
+
+        # Add the policy to the indices
+        client.add_ism_policy_to_index(policies.ISM_ID_HISTORY, policies.INDEX_PATTERN_HISTORY)
+
+def setup_sessions_ism(spi_days: int, replicas: int, client: OpenSearchClient):
+    # Create the new policy template
+    policy = policies.get_sessions_ism_policy(spi_days, 0, replicas, policies.ISM_DEFAULT_MERGE_SEGMENTS)
```
Comment (awick) [inline manage_arkime/opensearch_interactions/ism_interactions.py:34]: Probably want this to be
policy = policies.get_sessions_ism_policy(1, spi_days -1, replicas, policies.ISM_DEFAULT_MERGE_SEGMENTS)

This way we keep in hot for 1 day and then do a force merge.

File: manage_arkime/opensearch_interactions/ism_policies.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                    "index_patterns": [
+                        INDEX_PATTERN_HISTORY
+                    ],
+                    "priority": 95
+                }
+            ]
+        }
+    }
+
+ISM_ID_SESSIONS="sessions"
```
[the remark this answers is not in the document]
Comment (chelma) [reply manage_arkime/opensearch_interactions/ism_policies.py:43]: Guess I misread the perl code.  Will do.

File: manage_arkime/opensearch_interactions/ism_policies.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
+from typing import Dict
+
+ISM_ID_HISTORY="history"
```
[the remark this answers is not in the document]
Comment (chelma) [reply manage_arkime/opensearch_interactions/ism_policies.py:3]: Will do

File: manage_arkime/opensearch_interactions/ism_interactions.py (hunk 5 above)
Comment (chelma) [reply manage_arkime/opensearch_interactions/ism_interactions.py:34]: Sure, seems reasonable.  Will change.

**Comment by chelma** [comment]
Latest PR update should address all outstanding comments.

File: manage_arkime/opensearch_interactions/ism_policies.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (awick) [reply manage_arkime/opensearch_interactions/ism_policies.py]: perl is amazing! ;)

**Review by awick — APPROVED** [review]
LGTM

**Pull request merged**